cloudinit/net/network_manager.py

```diff
@@ -438,7 +461,7 @@ def render_network_state(
         # Now render the actual interface configuration
         for iface in network_state.iter_interfaces():
             conn = self.connections[iface["name"]]
-            conn.render_interface(iface, self)
+            conn.render_interface(iface, self, network_state)
```

Comment:
Rather than passing in `network_state` on this line, could we instead loop through the DNS entries and call `conn._add_nameserver()` after this line? We could remove the leading `_` so it's not a private method.

Same for adding the dns search. I think after a rebase that should work without doing the additional address checking.

Reply:
Not sure, it will become ugly either way I fear. Somehow we will need to check if there are static addresses and if yes, if there are no DNS options set already we need to set them. There are probably two options here: 

 * Do this in `render_network_state` which requires basically redoing what you already did (partially) in `render_interface`.
 * Add a method to the connection ala `finalize(network_state)` which then checks the `.config` and if it has addresses but no DNS yet it would apply the top-top-top-level DNS stuff.

No strong feelings and both will probably end up ugly, my gut tells me that a `finalize` option might maybe be nicer (slightly :D)